### PR TITLE
Load pre-trained models from HF Hub without unnecessary allocations

### PR DESCRIPTION
**Pull request by shadeMe**
## Description

This PR prevents the default initialization of models whose parameters will be replaced with those from HuggingFace Hub immediately. Furthermore, the loading of the serialized weights is performed in a zero-copy fashion when possible.

### Types of change

Enhancement

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: curated_transformers/models/embeddings.py [hunk 1]
```diff
@@ -48,14 +64,22 @@ def __init__(self, width: int, *, seq_len: int = 512, base: int = 10000):
...
+        :param device: Device on which the module is to be initialized.
+        """
         super().__init__()
 
         if width % 2:
             raise ValueError(f"Width of rotary embeddings must be even, was: {width}")
 
+        # Ignore allocations on the meta device as we don't persist our buffer,
+        # i.e., we don't expect the backing tensor to be replaced with pretrained weights.
+        if device == torch.device("meta"):
```
Comment (danieldk) [inline curated_transformers/models/embeddings.py:77]:
```suggestion
        if device.type == "meta":
```

File: curated_transformers/models/loading_util.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+    size_mismatch_msg = (
+        "Expected the size of the replacement for `{}` to be {}, but got {}"
+    )
+
+    if is_parameter:
+        old_param = module._parameters[tensor_name]
+        assert old_param is not None
+        if old_param.shape != tensor.shape:
+            raise ValueError(
```
Comment (danieldk) [inline curated_transformers/models/loading_util.py:38]: Do we also need to check that the dtype matches?

File: curated_transformers/models/loading_util.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                size_mismatch_msg.format(tensor_name, old_param.shape, tensor.shape)
+            )
+        new_param = Parameter(tensor, requires_grad=old_param.requires_grad)
+        module._parameters[tensor_name] = new_param
+    else:
+        old_buffer = module._buffers[tensor_name]
+        assert old_buffer is not None
+        if old_buffer.shape != tensor.shape:
```
Comment (danieldk) [inline curated_transformers/models/loading_util.py:46]: Same here (dtype).

File: curated_transformers/models/loading_util.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+                raise ValueError(f"Key `{name}` not found in module `{prefix}`")
+            replacement = state_dict[key]
+            _emplace_module_tensor(module, name, replacement, device=device)
+
+    def traverse(module: Module, state_dict: Mapping[str, torch.Tensor], prefix: str):
+        emplace(module, state_dict, prefix)
+
+        for name, child in module._modules.items():
+            if child is not None:
+                child_prefix = prefix + name + "."
```
Comment (danieldk) [inline curated_transformers/models/loading_util.py:97]:
```suggestion
                child_prefix = f"{prefix}{name}."
```

File: curated_transformers/models/loading_util.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+                child_state_dict = {
+                    k: v for k, v in state_dict.items() if k.startswith(child_prefix)
+                }
+                traverse(child, child_state_dict, child_prefix)
+
+    traverse(module, state_dict, "")
+    del traverse
+    del emplace
```
Comment (danieldk) [inline curated_transformers/models/loading_util.py:105]: Why are these needed? I could understand the deletion of `emplace` immediately after the `emplace` call (to avoid keeping it in memory while traversing, but doesn't do the interpreter the necessary analysis?)

**Review by danieldk — COMMENTED** [review]
Only some small comments/questions.

File: curated_transformers/models/loading_util.py (hunk 2 above)
Comment (shadeMe) [reply curated_transformers/models/loading_util.py:38]: Good question. It's probably a good idea, but once implement quantized weight loading, the types won't match anyway. So, I think we can dispense with it?

File: curated_transformers/models/loading_util.py (hunk 5 above)
Comment (shadeMe) [reply curated_transformers/models/loading_util.py:105]: I was just following [what `Module.load_state_dict` does](https://pytorch.org/docs/stable/_modules/torch/nn/modules/module.html#Module.load_state_dict), to be honest. I assumed it was to make sure that the `state_dict` wouldn't stick around, but it wouldn't make a difference in our case.

File: curated_transformers/models/loading_util.py (hunk 2 above)
Comment (shadeMe) [reply curated_transformers/models/loading_util.py]: Added the check.

File: curated_transformers/models/loading_util.py (hunk 5 above)
Comment (shadeMe) [reply curated_transformers/models/loading_util.py]: Rewrote the function to use a queue instead (and added some extra sanity checks).

**Review by danieldk — APPROVED** [review]

File: curated_transformers/models/loading_util.py (hunk 5 above)
Comment (danieldk) [reply curated_transformers/models/loading_util.py]: Much nicer!
